### PR TITLE
feat(kuma-cp) Support for udp

### LINKED ISSUE #239: Support DTLS and UDP routing
### Summary

Support DTLS or UDP routing? This is useful for game server hosting that uses kuma as a load balancer and for things like internet of things endpoints.

### Steps To Reproduce

Use kuma.

### Additional Details & Logs

- Version 2291f05

### PR DESCRIPTION
**Pull request by tharun208**
Signed-off-by: Tharun <rajendrantharun@live.com>

### Summary

Added support for udp protocol.

### Issues resolved

Fix #239 

### Documentation

- [ ] Link to the website [documentation PR](https://github.com/kumahq/kuma-website/pull/XXX)

### CODE REVIEW

File: pkg/xds/generator/outbound_proxy_generator.go [hunk 1]
```diff
@@ -101,6 +101,10 @@ func (_ OutboundProxyGenerator) generateLDS(proxy *model.Proxy, subsets []envoy_
 				Configure(envoy_listeners.Tracing(proxy.TracingBackend)).
 				Configure(envoy_listeners.HttpAccessLog(meshName, envoy_listeners.TrafficDirectionOutbound, sourceService, serviceName, proxy.Logs[serviceName], proxy)).
 				Configure(envoy_listeners.HttpOutboundRoute(envoy_names.GetOutboundRouteName(serviceName)))
+		case mesh_core.ProtocolUDP:
+			filterChainBuilder.
+				Configure(envoy_listeners.UdpProxy(serviceName, subsets[0])).
```
Comment (jakubdyszkiewicz) [inline pkg/xds/generator/outbound_proxy_generator.go:106]: Please add a comment here why we are choosing only the first subset (lack of support of subset loadbalancing for UDPProxy in Envoy)

**Comment by nickolaev** [comment]
@tharun208 can I suggest `netcat -ulp 12345` as a simple UDP server and then `netcat -cu <service> 12345` to connect. Also interesting is this going to work cross zones?

**Comment by tharun208** [comment]
> @tharun208 can I suggest `netcat -ulp 12345` as a simple UDP server and then `netcat -cu <service> 12345` to connect. Also interesting is this going to work cross zones?

sure @nickolaev

**Comment by nickolaev** [comment]
@tharun208 will you please merge master?

**Comment by nickolaev** [comment]
@tharun208 I would like to revamp this one before 1.0.0 is out. Can you please merge master again?
@jakubdyszkiewicz do you see any problems with this one?

File: pkg/xds/envoy/listeners/inbound_listener_configurer.go [hunk 2]
```diff
@@ -3,33 +3,43 @@ package listeners
 import (
 	v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
 	envoy_core "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	mesh_core "github.com/kumahq/kuma/pkg/core/resources/apis/mesh"
 )
 
-func InboundListener(listenerName string, address string, port uint32, isUDP bool) ListenerBuilderOpt {
+func InboundListener(listenerName string, protocol mesh_core.Protocol, address string, port uint32) ListenerBuilderOpt {
```
Comment (tharun208) [inline pkg/xds/envoy/listeners/inbound_listener_configurer.go:9]: 👍

**Comment by ktkfree** [comment]
hello.

I am waiting this feature.
any plans for this feature to be merged in master?

thanks.

**Comment by nickolaev** [comment]
@Mergifyio update

**Comment by tharun208** [comment]
finally, I found the issue and the envoy accepted the configuration :). The next step is to apply different policies and check everything working fine

**Comment by subnetmarco** [comment]
This PR has been open for a while, what is the current status? @tharun208

**Comment by tharun208** [comment]
> This PR has been open for a while, what is the current status? @tharun208

@subnetmarco the current status is the envoy listeners are generated for UDP protocol and I was able to see dp has a valid configuration without any rejections. I will test with a UDP client/server and apply some policies.

**Comment by subnetmarco** [comment]
I also see lots of conflicts to be resolved. Also perhaps it’s worth to do a chat with @jakubdyszkiewicz to determine how we can integrate this with the Envoy API v3 support he is building.

**Comment by nickolaev** [comment]
@mergifyio update

File: pkg/xds/envoy/listeners/v2/inbound_listener_configurer.go [hunk 3]
```diff
@@ -3,21 +3,32 @@ package v2
...
 func (c *InboundListenerConfigurer) Configure(l *v2.Listener) error {
 	l.Name = c.ListenerName
 	l.TrafficDirection = envoy_core.TrafficDirection_INBOUND
+	var listenerProtocol envoy_core.SocketAddress_Protocol
+	switch c.Protocol {
+	case mesh_core.ProtocolUDP:
+		listenerProtocol = envoy_core.SocketAddress_UDP
+		l.ReusePort = true
+	case mesh_core.ProtocolTCP:
+		listenerProtocol = envoy_core.SocketAddress_TCP
```
Comment (lahabana) [inline pkg/xds/envoy/listeners/v2/inbound_listener_configurer.go:26]: Nit: Maybe return an error if it's neither TCP nor UDP to make configuration/dev errors more obvious to users (atm this falls back to TCP which is ok 99% but can lead to hard to trackdown bugs).

**Review by lahabana — COMMENTED** [review]
This does a good part of the work I need for the DNSFilter support! Thanks!
2 small aesthetic comments:

- Fail fast when using non L4 protocols in ListenerConfigurers
- Maybe don't add ListenerProtocol parameter on tests that are only valid with TCP. IMO it just makes tests harder to follow

> BTW I'm not a maintainer so don't take this authoritatively :). I just need this PR merged for my work.

**Comment by tharun208** [comment]
@nickolaev could I get a review on this.

**Comment by schmurfy** [comment]
I just found out about kuma but for my usecase I need udp and tcp, was there anything preventing this pr from being merged ?

**Comment by lahabana** [comment]
This PR requires extra work and a serious rebase. I'm going to close it and we'll be able to reuse it when really working on #239 .

**Pull request closed without merging**